src/zkevm_specs/evm/execution/sar.py

```diff
@@ -0,0 +1,185 @@
+from ...util import (
+    FQ,
+    MAX_U64,
+    N_BYTES_U64,
+    RLC,
+    int_is_neg,
+    word_to_lo_hi,
+)
+from ..instruction import Instruction, Transition
+from ..typing import Sequence
+
+
+def sar(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    shift = instruction.stack_pop()
+    a = instruction.stack_pop()
+    b = instruction.stack_push()
+
+    (
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+    ) = gen_witness(instruction, shift, a)
+    check_witness(
+        instruction,
+        shift,
+        a,
+        b,
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        p_lo,
+        p_hi,
+        p_top,
+    )
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(3),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(1),
+    )
+
+
+def check_witness(
+    instruction: Instruction,
+    shift: RLC,
+    a: RLC,
+    b: RLC,
+    a64s: Sequence[FQ],
+    b64s: Sequence[FQ],
+    a64s_lo: Sequence[FQ],
+    a64s_hi: Sequence[FQ],
+    shf_div64: FQ,
+    shf_mod64: FQ,
+    p_lo: FQ,
+    p_hi: FQ,
+    p_top: FQ,
+):
+    is_neg, _ = instruction.compare(FQ(127), FQ(a.le_bytes[31]), 1)
+    shf_lo, shf_hi = instruction.word_to_lo_hi(shift)
+    shf_hi_is_zero = instruction.is_zero(shf_hi)
+
+    for idx in range(4):
+        offset = idx * N_BYTES_U64
+
+        # a64s constraint
+        instruction.constrain_equal(
+            a64s[idx],
+            instruction.bytes_to_fq(a.le_bytes[offset : offset + N_BYTES_U64]),
+        )
+
+        # b64s constraint
+        instruction.constrain_equal(
+            instruction.select(shf_hi_is_zero, b64s[idx], is_neg * MAX_U64),
```

Comment:
shouldn't here be the case that `b.le_bytes[offset : offset + N_BYTES_U64]` always equals to `b64s[idx]` even when `shf_hi_is_zero` is 0?

Reply:
Yes. I see. I fixed to only calculate `b64s` when `shift < 256` (`shf_hi == 0 and shf_lo // 64 < 4`). It could reduce useless calculation for `b64s`. Thanks.

In `gen_witness`, I fixed to only calculate `b64s` when `shf_hi == 0 and shf_div64 < 4`, otherwise use initialized values. And updated constraints in `check_witness`.

Also updated Markdown doc, and added below two test cases (`shf_hi != 0 and shf_lo // 64 < 4`):
```
shift = 0x80000... + 8, a = 0x1234
shift = 0x80000... + 8, a = 0x80000... + 0x1234
```